src/rendergraph/common/rendergraph/texture.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <QImage>
+#include <memory>
+
+#include "backend/texture.h"
+
+namespace rendergraph {
+class Context;
+class Texture;
+} // namespace rendergraph
+
+class rendergraph::Texture {
+  public:
+    Texture(Context& context, const QImage& image);
+
+    backend::Texture* backendTexture() const {
+        return m_pTexture.get();
+    }
+
+  private:
+    const std::unique_ptr<backend::Texture> m_pTexture{};
```

Comment:
should the pointer or the texture be const? currently its just the pointer afaict.

Reply:
The pointer. Its lifetime is tight to its owner.

Reply:
right, but thats not implying const. As already pointed out, this inhibits moves (and due to the unique_ptr) also copies. That's artificially limiting and could cause consumers to wrap this class in yet another unique_ptr to work around the immovability (even though there is no runtime polymorphism which would require it).

Reply:
But I am already using Texture wrapped in a std::unique_ptr<Texture> , and I don't see that as a problem.

See for example:

```
rendergraph::TextureMaterial
   void setTexture(std::unique_ptr<Texture> texture) 
```

This is the typical use of Texture, create it with ```std::make_unique<rendergraph::Texture>(context, img))``` and pass the ownership to the material.

(Now, possibly it should be std::shared_ptr<Texture>, but not now)

Maybe I am missing your point, but I don't see any advantage of moving Texture instances, I think it would make the code much less clear than using std::unique_ptr<Texture> , with it's obvious move semantics.

(On a side note, the backend::Texture *has*  to be a pointer. In the case of OpenGL we could simply store the QOpenGLTexture itself, but in the case of scenegraph, the QSGTexture* *has* to be created by the QQuickWindow (which is why we need to pass the context to the rendergraph::Texture constructor, unused in the case of the opengl backend)